### PR TITLE
Audit Log Plugin

### PR DESCRIPTION
**Pull request by aydreeihn**
This new plugin tracks what Users and Agents do in the helpdesk.

Check out the documentation here:
https://github.com/osTicket/docs/blob/195812993b954d6bb283d33e824d3f42d2b16132/Plugins/Audits.rst

You'll need to install the plugin using the attached file as well. For help installing plugins, here's a tutorial:
https://youtu.be/Olh1z-RNEhQ

Note: You will have to change the file extension from .txt to .phar (Git wouldn't let me upload it as a phar file)
[audit.txt](https://github.com/osTicket/osTicket/files/3622341/audit.txt)

### CODE REVIEW

**Comment by johnkalangha** [comment]
hi, what version of osticket need to run this plugin, i've tried this in 1.14RC1 but nothing happens thankyou

**Comment by aydreeihn** [comment]
@johnkalangha You should be able to run it using 1.14RC1. Did you make sure to add the .phar file to your plugin folder as well as enable the audit log plugin after adding it?

**Comment by johnkalangha** [comment]
yes maam i already change the file extension from txt to phar and i already enable it to my osticket, but when i tried to see the changes in agent or user side, i dont see the audit tab like in the documentation. do i need to do something or what im thinking is right? Thank you

**Comment by aydreeihn** [comment]
If all of those steps were done, it should work just fine. I will need to see some screenshots to help you out. Can you go to Admin Panel | Manage | Plugins and take a screenshot of what you see there?

**Comment by johnkalangha** [comment]
hi maam this is my screenshot

Admin Panel | Manage | Plugins
image

Users 
image

Ticket
image

**Comment by aydreeihn** [comment]
@johnkalangha It looks like you might not have merged the code from this Pull Request into the osTicket codebase. That should solve the problem for you.

**Comment by philbertphotos** [comment]
Kinda slick

**Comment by philbertphotos** [comment]
not really a true plugin if you hack the codebase for it to work ...but sweet modification none the less.

**Comment by JediKev** [comment]
@philbertphotos

Umm...I’m not sure what you actually mean by this. She is a core dev and this is the official plugin (well not all the way official as it’s not packaged yet). It’s actually really clean code and an amazing feature. Yes she had to modify some of core but plugins are allowed to do that. Especially if it enhances core.

Cheers.

File: include/class.dept.php [hunk 1]
```diff
@@ -813,6 +823,24 @@ function update($vars, &$errors) {
         if ($errors)
             return false;
 
+        $vars['disable_auto_claim'] = isset($vars['disable_auto_claim']) ? 1 : 0;
+        if (PluginManager::auditPlugin() && $this->getId()) {
```
Comment (protich) [inline include/class.dept.php]: Is this necessary?

File: include/class.template.php [hunk 2]
```diff
@@ -455,6 +455,13 @@ function save($id, $vars, &$errors) {
 
         if($errors) return false;
 
+        foreach ($vars as $key => $value) {
+            if ($id && PluginManager::auditPlugin() && isset($this->ht[$key]) && ($this->ht[$key] != $value)) {
```
Comment (protich) [inline include/class.template.php]: PluginManager::auditPlugin()

File: scp/orgs.php [hunk 3]
```diff
@@ -72,9 +72,13 @@
             $count = 0;
             switch (strtolower($_POST['do'])) {
             case 'delete':
-                foreach ($orgs as $O)
-                    if ($O->delete())
-                        $count++;
+                foreach ($orgs as $O) {
+                  $type = array('type' => 'deleted');
+                  Signal::send('object.deleted', $O, $type);
```
Comment (protich) [inline scp/orgs.php]: Why not send the signal in delete routine?

File: scp/staff.php [hunk 4]
```diff
@@ -101,8 +103,12 @@
                     case 'delete':
                         $i = 0;
                         foreach($members as $s) {
-                            if ($s->staff_id != $thisstaff->getId() && $s->delete())
-                                $i++;
+                            if ($s->staff_id != $thisstaff->getId()) {
+                              $type = array('type' => 'deleted');
+                              Signal::send('object.deleted', $s, $type);
+                              $s->delete();
```
Comment (protich) [inline scp/staff.php:107]: Signal getting sent before the actual deletion happens - also see my other comment about sending signal in delete function itself

**Review by protich — COMMENTED** [review]
I think you're doing too much explicit inspection to detect edits on the object. ORM objects already track edits, using `dirty`,  allowing you to do something like... 

```php
if ($obj->dirty) 
  Signal::send(.... 
```

**Comment by philbertphotos** [comment]
@JediKev I had no idea that all those core changes would be part of part of the official release of osticket ... Weill dam.. guess its not a hack, Still Great plugin none the less love what I see

**Comment by ntozier** [comment]
Side note: shouldn't this actually be in this repo?
https://github.com/osTicket/osTicket-plugins

**Comment by JediKev** [comment]
@ntozier

It is:
- https://github.com/osTicket/osTicket-plugins/pull/161

She needed core mods as well in tangent with the plugin code.

Cheers.

**Comment by ntozier** [comment]
ah gotcha.

**Comment by carranzamarioagustin** [comment]
> 
> 
> @johnkalangha It looks like you might not have merged the code from this Pull Request into the osTicket codebase. That should solve the problem for you.

I'm really new to github, could you help me about how to merge this plugin to the code in my server?

**Comment by aydreeihn** [comment]
@Tekkness 
The commands below will create a new branch with this feature on it from your current branch:
git checkout -b aydreeihn-feature/audit-logs develop-next
git pull https://github.com/aydreeihn/osTicket.git feature/audit-logs

We also have a video tutorial about how to keep your codebase up to date with GitHub here:
https://youtu.be/0R4xu118LzM

**Comment by carranzamarioagustin** [comment]
> 
> 
> @Tekkness
> The commands below will create a new branch with this feature on it from your current branch:
> git checkout -b aydreeihn-feature/audit-logs develop-next
> git pull https://github.com/aydreeihn/osTicket.git feature/audit-logs
> 
> We also have a video tutorial about how to keep your codebase up to date with GitHub here:
> https://youtu.be/0R4xu118LzM

Since I didnt use git to install OsTicket (downloaded the zip and unzipped), im getting the "not a git repository" error.

Would getting the zip and replacing all work?

**Comment by aydreeihn** [comment]
@Tekkness I think you would have to clone the repository to a new folder and then reconfigure your include/ost-config.php file to point to your database within the cloned folder instead of the one you're currently using that isn't a git repo.

**Comment by johnkalangha** [comment]
Hi when i using audit logs in dashboard of admins, this is the error logs in osticket version 1.14.1

[Fri Nov 22 09:53:57.990209 2019] [php7:warn] [pid 6116:tid 1924] [client ::1:59805] PHP Warning:  require_once(phar:///C:/xampp/htdocs/osticket/include//plugins/audit.phar/class.audit.php): failed to open stream: phar error: invalid url or non-existent phar &quot;phar:///C:/xampp/htdocs/osticket/include//plugins/audit.phar/class.audit.php&quot; in C:\\xampp\\htdocs\\osticket\\scp\\audits.php on line 19, referer: http://localhost/osticket/scp/logs.php
[Fri Nov 22 09:53:57.990209 2019] [php7:error] [pid 6116:tid 1924] [client ::1:59805] PHP Fatal error:  require_once(): Failed opening required

**Comment by johnkalangha** [comment]
i just want to ask if did you remove the settings in ROLES for audit log plugin? i think this is a good idea to minimize the agent privilege the view the audit logs thanks

**Comment by JediKev** [comment]
@johnkalangha

Looks like you have permission/ownership issues. Please correct these and retest.

It also looks there is an incorrect url but I don’t know if it’s the phar itself or your environment.

Cheers.

**Comment by johnkalangha** [comment]
@JediKev  i dont know what file do i need to correct in permission, this is a new fresh installed.. audit logs seems running when i use in specific ticket in ticket Tab (in more settings) but when i check in specific ticket using USER tab, still there's nothing i can see, and there is another error in error log

**Comment by JediKev** [comment]
@johnkalangha

>[Fri Nov 22 09:53:57.990209 2019] [php7:error] [pid 6116:tid 1924] [client ::1:59805] PHP Fatal error: require_once(): Failed opening required
 
The error message is not complete so I’m not sure either.

> "phar:///C:/xampp/htdocs/osticket/include//plugins/audit.phar/class.audit.php" in C:\xampp\htdocs\osticket\scp\audits.php on line 19

Looks like you may need to remove the extra slash before `/plugin` on line 19 and 21.
- https://github.com/osTicket/osTicket/blob/develop-next/scp/audits.php#L21
- https://github.com/osTicket/osTicket/blob/develop-next/scp/audits.php#L19

Cheers.

**Comment by johnkalangha** [comment]
here sir 

&quot;phar:///C:/xampp/htdocs/osticket/include//plugins/audit.phar/templates/agent-audit.tmpl.php&quot; in phar://C:/xampp/htdocs/osticket/include/plugins/audit.phar/audit.php on line 64, referer: http://localhost/osticket/scp/staff.php
[Fri Nov 22 10:05:04.668761 2019] [php7:warn] [pid 6116:tid 1924] [client ::1:60133] PHP Warning:  include(): Failed opening 'phar:///C:/xampp/htdocs/osticket/include//plugins/audit.phar/templates/agent-audit.tmpl.php' for inclusion (include_path='./;C:/xampp/htdocs/osticket/include/;C:/xampp/htdocs/osticket/include/pear/') in phar://C:/xampp/htdocs/osticket/include/plugins/audit.phar/audit.php on line 64, referer: http://localhost/osticket/scp/staff.php
[Fri Nov 22 10:05:12.247966 2019] [php7:warn] [pid 6116:tid 1924] [client ::1:60136] PHP Warning:  include(phar:///C:/xampp/htdocs/osticket/include//plugins/audit.phar/templates/agent-audit.tmpl.php): failed to open stream: phar error: invalid url or non-existent phar &quot;phar:///C:/xampp/htdocs/osticket/include//plugins/audit.phar/templates/agent-audit.tmpl.php&quot; in phar://C:/xampp/htdocs/osticket/include/plugins/audit.phar/audit.php on line 64, referer: http://localhost/osticket/scp/staff.php?id=1
[Fri Nov 22 10:05:12.248969 2019] [php7:warn] [pid 6116:tid 1924] [client ::1:60136] PHP Warning:  include(): Failed opening 'phar:///C:/xampp/htdocs/osticket/include//plugins/audit.phar/templates/agent-audit.tmpl.php' for inclusion (include_path='./;C:/xampp/htdocs/osticket/include/;C:/xampp/htdocs/osticket/include/pear/') in phar://C:/xampp/htdocs/osticket/include/plugins/audit.phar/audit.php on line 64, referer: http://localhost/osticket/scp/staff.php?id=1

when i opened in  agent audit logs.

**Comment by JediKev** [comment]
@johnkalangha

Refer to my comment above. Looks like your PHP can’t find the file with the extra slash in the path (which is weird as with phar it's typically smart enough to figure it out). Just read the errors and find what file and what line it says and remove the extra slashes.

We will take a look at this in the morning.

Cheers.

**Comment by johnkalangha** [comment]
noted sir, thank you very much

**Comment by k0jdd** [comment]
Another bug to look into in PHP 7.3 i get the following error related to ticket audits (done via ajax)

`[04-Apr-2020 21:34:58 UTC] PHP Warning: include(): Failed opening 'templates/ticket-audit.tmpl.php' for inclusion (include_path='./:/home/netlogger/www/support/include/:/home/netlogger/www/support/include/pear/') in phar:///home/netlogger/www/support/include/plugins/audit.phar/audit.php on line 80`

the corresponding code is as follow:
  

>      // Ajax View Ticket Audit
>         Signal::connect('ajax.scp', function($dispatcher) {
>             $dispatcher->append(
>                 url_get('^/audit/ticket/(?P<id>\d+)/view$', function($ticketId) {
>                     global $thisstaff;
> 
>                     $row = Ticket::objects()->filter(array('ticket_id'=>$ticketId))->values_flat('number')->first();
>                     if (!$row)
>                         Http::response(404, 'No such ticket');
>                     if (!$thisstaff || !$thisstaff->isAdmin())
>                         Http::response(403, 'Contact your administrator');
> 
>                     include 'templates/ticket-audit.tmpl.php';
>                 })
>             );
>         });`

PHP 7.1 works fine

Just one more thing to look into

**Comment by aydreeihn** [comment]
@k0jdd Do you get this error when you click 'View Audit Log' from within a Ticket? Also, are you using a Windows or Linux environment? I don't see any errors in my Linux environment using PHP 7.3, but we have seen some differences between the two. I've made a few corrections on the plugin since we originally released it, and I wonder if maybe this isn't a bug anymore. Would you mind trying the attached file to see if you get the same error? Just change the file extension from .txt to .phar once you download it.

[audit.txt](https://github.com/osTicket/osTicket/files/4438871/audit.txt)

**Comment by k0jdd** [comment]
@aydreeihn   Yes, I get this error when I  click 'View Audit Log' from within a Ticket

I loaded the new plugin provided by you and the error persists on PHP7.3 - No error on PHP7.1
Same warning message on line 80 as before.  I suspect the issue is that PHP7.3 is not always understanding where to look for the include files (i.e. not searching in the .phar)  

Line 80 has an relative pathname for the include - which PHP should interpret to be the current directory (or in the include path as a last resort).  The question then is what does PHP  7.3 believe  the current directory to be? Is it plugins/ , or plugins/audit.phar/ , or something else? Obviously PHP7.1 knows where to look.  

Info about my PHP versions are at these links: 
v7.1 (working) http://support.netlogger.org/php_info.php
V7.3 (not working) http://www.netlogger.org/php_info.php

The underlying OS is Debian 10 "Buster" Linux KVM infrastructure 
I do not control the OS or PHP engines - the hosting provider does that.

Is there anything else I can tell you? I am willing to test more if you think of something to try.

I am getting along fine on PHP7.1 although there are significant performance gains to be had with PHP7.3

**Comment by k0jdd** [comment]
Bump

**Comment by JediKev** [comment]
@k0jdd

Have you tried with the latest release of v1.14.2? We published it yesterday and it includes a ton of bug fixes. Its available for download [here](https://osticket.com/download).

**Edit:**
We should be releasing new builds of plugins and language packs tomorrow so stay tuned for that as well.

Cheers.

**Comment by k0jdd** [comment]
@JediKev @aydreeihn 

 I have upgraded to 1.14.2. No change in behavior when using php 7.3   

The error points to a failure to include a script due to the INCLUDE_PATH not including the directory where the script is located. 

Curious as to why you think that upgrading would have made a difference unless someone has actually researched the issue and come up with a fix.  That might be a good next step. 

Once again I am willing to help anyway that I can...

**Comment by JediKev** [comment]
@k0jdd 

Yes, we included some Audit Log bug fixes in the core codebase as well as the Plugin itself. It looks like you'll need to download and install a new build of the plugin as we included path fixes with the new build. You can download the new build below:
- https://osticket.com/download

Cheers.

**Comment by k0jdd** [comment]
No change.

`[14-May-2020 14:59:55 UTC] PHP Warning:  include(): Failed opening 'templates/ticket-audit.tmpl.php' for inclusion (include_path='./:/home/netlogger/www/support/include/:/home/netlogger/www/support/include/pear/') in phar:///home/netlogger/www/support/include/plugins/audit.phar/audit.php on line 80
`

**Comment by JediKev** [comment]
@k0jdd 

We will have to look into this and get back to you as I personally cannot replicate this in a Linux environment.

Cheers.

**Comment by k0jdd** [comment]
@JediKev 

Thanks for the responses.

Maybe you can try dumping the include_path on your Linux server and see what it contains. You can see mine in my previous post.

Your codes reads: 
`include 'templates/ticket-audit.tmpl.php';` 

Why are you including the templates contained in the phar  with a relative path? I suspect that is no longer allowed in PHP7.3 (or possibly it's broken)

Here is what the Documentation says to do.
Reference:  https://www.php.net/manual/en/intro.phar.php

I suspect you may need to construct an explicit path for your include statement here such as"
`<?php
 include 'phar:///full/path/to/include/plugins/audit.phar/templates/ticket-audit.tmpl.php';
 ?>`

**Comment by JediKev** [comment]
@k0jdd 

My `include_path` is the exact same as yours (`./:/path/to/v1.14.2/include/:/path/to/v1.14.2/include/pear/`) plus I'm running PHP 7.4 so I don't know what the difference could be between our two environments. 🤷🏼‍♂️

Could you try this build of the plugin and let me know? (tested and working in all of my Linux and Windows environments with PHP 5.6-7.4 and MySQL 5.5.52-8.0.17)
[audit.zip](https://github.com/osTicket/osTicket/files/4630893/audit.zip)

**Note:**
In the PHP Docs, it says using relative paths is still accepted in the PHP 7.x series.

Cheers.

**Comment by k0jdd** [comment]
@JediKev @aydreeihn 

Yes, that build works on PHP 7.3.1  -- thank you for following up.

Clearly there is something different about PHP7.3 (versus other versions).   

It understands the following:
` include sprintf('phar://%s/plugins/audit.phar/templates/ticket-audit.tmpl.php', INCLUDE_DIR);`

But apparently does not understand this:
`include 'templates/ticket-audit.tmpl.php';`

**Pull request merged**